### ISSUE TITLE
Function with nested quadgk calls takes 30 minutes to compile

### ISSUE THREAD

**Issue by chrstphrbrns** [JuliaLang/julia#45769]
```julia
julia> using QuadGK

julia> f(a,b,c) = quadgk(x -> first(quadgk(y -> first(quadgk(z -> cos(x * y * z), -c/2, c)), -b/2, b)), 0, a)
f (generic function with 1 method)

julia> @time f(5,5,5)
1967.236909 seconds (6.86 G allocations: 414.765 GiB, 6.38% gc time, 99.85% compilation time)
(29.609263448810637, 1.948074789259735e-7)
```
```julia
julia> versioninfo()
Julia Version 1.9.0-DEV.822
Commit 782458c221* (2022-06-20 19:04 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 32 × AMD Ryzen 9 3950X 16-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.3 (ORCJIT, znver2)
  Threads: 32 on 32 virtual cores
Environment:
  LD_PRELOAD = /lib/x86_64-linux-gnu/libc_malloc_debug.so.0
  LD_LIBRARY_PATH = :/usr/local/cuda-11.2/lib64:/opt/TensorRT-7.0.0.11/lib/:/opt/libcutensor/lib/10.2
  JULIA_NUM_THREADS = 32
  JULIA_EDITOR = vim
```
In ```1.7.3``` this takes ~2 minutes
```julia
julia> VERSION
v"1.7.3"

julia> using QuadGK

julia> f(a,b,c) = quadgk(x -> first(quadgk(y -> first(quadgk(z -> cos(x * y * z), -c/2, c)), -b/2, b)), 0, a)
f (generic function with 1 method)

julia> @time f(5,5,5)
132.260530 seconds (606.12 M allocations: 34.966 GiB, 4.05% gc time, 97.63% compilation time)
(29.609263448810637, 1.948074789259735e-7)
```

**Comment by oscardssmith** [comment]
That said, 2 minutes on 1.7 is also pretty bad. Why is the compiler so mad at this code?

**Comment by JeffBezanson** [comment]
Hmm, on 1.8 we have
```
METHOD_MATCH              : 91.07 %   1718600306698
```

**Comment by chrstphrbrns** [comment]
If you change all of the lower integration bounds to 0, then it only takes 30 seconds to compile in `1.9` (vs 50 seconds in `1.7.3`)

**Comment by oscardssmith** [comment]
that's really weird.

**Comment by JeffBezanson** [comment]
What do you know, I won the bisect: 2c0e1bccb166d6b51ac09793acbd3fa09e2aa917
Makes sense, since it does more inference, but surprising that it does that much.

**Comment by JeffBezanson** [comment]
If the integration lower bounds are 0, then all the trailing arguments to quadgk are integers, and it has a `T...` method, so that probably explains that.

**Comment by vtjnash** [comment]
Solved
```
julia> @time f(5,5,5)
  5.005164 seconds (18.48 M allocations: 1.066 GiB, 2.33% gc time, 75.48% compilation time)
(29.609263448810633, 1.948074784818843e-7)
```

**Comment by oscardssmith** [comment]
Specifically, this has been solved sometime between 1.10 and master. On 1.10 I get 45 seconds, vs 5 for 1.11.

**Comment by Youjack** [comment]
I find that if I use the internal function `QuadGK.do_quadgk` instead of `quadgk`, the compilation time can be greatly reduced (allocations are also greatly reduced). `quadgk` itself is calling `QuadGK.do_quadgk` internally, see https://github.com/JuliaMath/QuadGK.jl/blob/89a3b8fbeb8f6861617e39e510647bd8111dc584/src/api.jl#L80-L85.

First of all, `@time f(5,5,5)` actually takes more than 20 seconds on Apple M1 Max and AMD 7940HS (counted using a stopwatch or Jupyter notebook in vscode), instead of ~5 seconds that `@time` tells.

Using `quadgk` (~23 seconds on stopwatch):
```julia
julia> using QuadGK

julia> f(a,b,c) = quadgk(x -> first(quadgk(y -> first(quadgk(z -> cos(x * y * z), -c/2, c)), -b/2, b)), 0, a)
f (generic function with 1 method)

julia> @time f(5,5,5)
  4.715462 seconds (50.72 M allocations: 2.567 GiB, 2.05% gc time, 76.07% compilation time)
(29.609263448810633, 1.948074784818843e-7)
```

Restart Julia and define a function `quad` that wraps `QuadGK.do_quadgk` (~2 seconds on stopwatch):
```julia
julia> using QuadGK

julia> function quad(f, a, b)
           QuadGK.do_quadgk(f, (Float64(a),Float64(b)), 7, nothing, nothing, 10^7, QuadGK.norm, nothing, nothing)
       end
quad (generic function with 1 method)

julia> f(a,b,c) = quad(x -> first(quad(y -> first(quad(z -> cos(x * y * z), -c/2, c)), -b/2, b)), 0, a)
f (generic function with 1 method)

julia> @time f(5,5,5)
  1.708222 seconds (3.81 M allocations: 409.958 MiB, 0.92% gc time, 33.96% compilation time)
(29.609263448810633, 1.948074784818843e-7)
```

```julia
julia> versioninfo()
Julia Version 1.11.6
Commit 9615af0f269 (2025-07-09 12:58 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin24.0.0)
  CPU: 10 × Apple M1 Max
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, apple-m1)
Threads: 8 default, 0 interactive, 4 GC (on 8 virtual cores)
Environment:
  JULIA_NUM_THREADS = auto
```
